File: CS/AsyncDataLoading/Core/AsynchronousCollectionFactory.cs
using System;
using System.Collections;
using System.Collections.Generic;
using System.Collections.Specialized;
using System.Diagnostics;
using System.Threading;
using System.ComponentModel;

namespace AsyncDataLoading {
    public enum RequestDataMode { OnDemand, InBackgroundThread }
    public static class AsynchronousCollectionFactory {
        public static AsynchronousCollection<T> CreateCollection<T>(RequestCount requestCount, RequestData requestData, RequestDataMode requestDataMode) {
            if(requestDataMode == RequestDataMode.OnDemand) return new AsynchronousCollection<T>(requestCount, requestData);
            else return new AsynchronousCollection2<T>(requestCount, requestData);
        }
        public static AsynchronousCollection<T> CreateCollection<T>(RequestCount requestCount, RequestData requestData, AsynchronousCollectionSettings settings) {
            return CreateCollection<T>(requestCount, requestData, null, settings);
        }
        public static AsynchronousCollection<T> CreateCollection<T>(RequestCount requestCount, RequestData requestData, SubmitChanges submitChanges, AsynchronousCollectionSettings settings) {
            if(settings.GetType() == typeof(AsynchronousCollection2Settings))
                return new AsynchronousCollection2<T>(requestCount, requestData, submitChanges, (AsynchronousCollection2Settings)settings);
            if(settings.GetType() == typeof(AsynchronousCollectionSettings))
                return new AsynchronousCollection<T>(requestCount, requestData, submitChanges, settings);
            return null;
        }
    }
}

File: VB/AsyncDataLoading/Core/AsynchronousCollectionFactory.vb
Imports System
Imports System.Collections
Imports System.Collections.Generic
Imports System.Collections.Specialized
Imports System.Diagnostics
Imports System.Threading
Imports System.ComponentModel

Namespace AsyncDataLoading
    Public Enum RequestDataMode
        OnDemand
        InBackgroundThread
    End Enum
    Public NotInheritable Class AsynchronousCollectionFactory

        Private Sub New()
        End Sub

        Public Shared Function CreateCollection(Of T)(ByVal requestCount As RequestCount, ByVal requestData As RequestData, ByVal requestDataMode As RequestDataMode) As AsynchronousCollection(Of T)
            If requestDataMode = AsyncDataLoading.RequestDataMode.OnDemand Then
                Return New AsynchronousCollection(Of T)(requestCount, requestData)
            Else
                Return New AsynchronousCollection2(Of T)(requestCount, requestData)
            End If
        End Function
        Public Shared Function CreateCollection(Of T)(ByVal requestCount As RequestCount, ByVal requestData As RequestData, ByVal settings As AsynchronousCollectionSettings) As AsynchronousCollection(Of T)
            Return CreateCollection(Of T)(requestCount, requestData, Nothing, settings)
        End Function
        Public Shared Function CreateCollection(Of T)(ByVal requestCount As RequestCount, ByVal requestData As RequestData, ByVal submitChanges As SubmitChanges, ByVal settings As AsynchronousCollectionSettings) As AsynchronousCollection(Of T)
            If settings.GetType() Is GetType(AsynchronousCollection2Settings) Then
                Return New AsynchronousCollection2(Of T)(requestCount, requestData, submitChanges, DirectCast(settings, AsynchronousCollection2Settings))
            End If
            If settings.GetType() Is GetType(AsynchronousCollectionSettings) Then
                Return New AsynchronousCollection(Of T)(requestCount, requestData, submitChanges, settings)
            End If
            Return Nothing
        End Function
    End Class
End Namespace
